### PR TITLE
Fix #2595 by skipping reflexive replacements

### LINKED ISSUE #2595: Cargo non-deterministically rebuilding and failing to override with [replace]
`cargo build --target i686-unknown-linux-gnu` with https://github.com/RustOS-Fork-Holding-Ground/RustOS/tree/3743e64e1db53e5822446c8c25ab7670cf590a75 seems to send Cargo haywire.

I use the `[replace]` section to substitute my own version of libraries with which explicitly depend on core. But dependencies will sometimes fail to be built because missing/duplicate core, indicating the `[replace]` did not work and it's either both linking the custom-built core and the one in the sysroot, or cannot find the one in the sysroot. Furthermore after building does succeed, it may seemingly at random try to rebuild dependencies again, in which case more `[replace]`-roulette.

Note that one does need the i686 sysroot installed to observe this---instead of sometimes succeeding, it will (as should be expected) make it to linking and then complain about missing compiler-rt.

Witnessed on the with the 2016-04-20 cargo and rust nighties.

### PR DESCRIPTION
**Pull request by Ericson2314**
If you know of something better than `source_id == source_id`, I'd gladly change it.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Could you explain a bit at a high level what happened here? I'm trying to decipher this from the commit itself but "Replace bug" isn't very helpful nor is the PR description...

**Comment by Ericson2314** [comment]
@alexcrichton Oh sorry! I thought I already edited it to take the top line of the last commit not the branch name.

**Comment by Ericson2314** [comment]
Ok so before this, if a dependency matches the replacement, it will be replaced with itself (at least according to the lockfile). Replaced packages don't have their dependencies listed, so while the reflexive replacement itself works, Cargo assume the package has zero deps.

In the example, local replaces `near` with itself, and `near` depends on `far`. Right now that test fails when `near` tries to extern `far`.

**Comment by alexcrichton** [comment]
Ok, thanks! That sounds like it'd definitely do the trick.

**Comment by Ericson2314** [comment]
NP! Again sorry about the awful PR title.

File: src/cargo/core/resolver/mod.rs [hunk 1]
```diff
@@ -785,46 +785,51 @@ impl<'a> Context<'a> {
...
-                if !spec.matches(summary.package_id()) {
-                    continue
-                }
-
-                if replace.is_some() {
-                    bail!("overlapping replacement specifications found:\n\n  \
-                           * {}\n  * {}\n\nboth specifications match: {}",
-                          matched_spec.unwrap(), spec, summary.package_id());
-                }
+            // get around lack of non-lexical lifetimes
```
Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: It's ok to remove comments like this

**Comment by alexcrichton** [comment]
@bors: r+ 3f655e54a45e130d45098ae3ad251774fc1b7179

Looks good to me, thanks @Ericson2314!

File: src/cargo/core/resolver/mod.rs (hunk 1 above)
Comment (Ericson2314) [inline src/cargo/core/resolver/mod.rs]: Should I force--push? (I wrote that, as opposed to move it around, to be clear.)

Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: If it bounces sure but otherwise it's fine

**Pull request merged**